examples/with-next-auth/pages/api/auth/[...nextauth].js

```diff
@@ -0,0 +1,105 @@
+import NextAuth from 'next-auth'
+import Providers from 'next-auth/providers'
+
+const options = {
+  site: process.env.SITE,
```

Comment:
Is there a way we can make this work without using an absolute URL here?

Reply:
@iaincollins

Reply:
Otherwise the instructions for the deployment would have to be more clear about how to add environment variables for production (due to `process.env.SITE` being `localhost`), and the deploy to Vercel button would have to be removed

Reply:
Yeah good call, I don't think this would work with a completely dynamic hostname / URL. You could put the assigned now.sh URL/a custom domain in the env variable after the first deploy though.

Reply:
It can be done with an absolute URL too, but that has to be explained in the readme. In the case of Vercel you would have to add the environment variable with the production domain in the environment variables section under settings.

Reply:
For security reasons an absolute URL must unfortunately be specified manually.

* It's possible to grab the hostname from the request object (e.g. `req.host`) but that is supplied by the client and can't be trusted.
* You can infer the URL of the instance but it's impossible to guess the intended canonical URL of an instance and that's important for OAuth integration to work

   e.g. callback URLs at `www.example.com` and `example.com` are not the same as far most OAuth providers are concerned - typically it has to match exactly whatever you have told them the URL of your site is.

Reply:
We _could_ maybe do something like:

`site: process.env.VERCEL_URL || 'http://localhost:3000'` 

But I'm not sure if `VERCEL_URL` contains a canonical / primary alias or if it's always an instance ID. Can anyone at Vercel advise if that's a good way to get what is "probably" the canonical hostname for a site?

Reply:
Okay so I've adjusted the `[...nextauth].js` file to contain `site: process.env.VERCEL_URL || 'http://localhost:3000'` in the latest commit

Reply:
To add to the discussion, I think most oauth providers don't support wildcard domains, so for instance when doing a preview deployment using next-auth with e.g. Google as an oauth provider, the unique url provided in `VERCEL_URL` will not work as a callback from Google unless explicitly added to the dashboard as far as I am aware.

I think when testing this I had to use something different than `VERCEL_URL` for `site` because even when merging to the default branch, `VERCEL_URL` had a unique url.

Reply:
Thank you for the clear answers 😌